### ISSUE TITLE
TestCaseSources initialization

### ISSUE THREAD

**Issue by ptr1120** [nunit/nunit#1256]
Thanks for the great work with NUnit.
I recently refactored all our UnitTests from MsTest to Nunit. We heavily use the parameterized Tests (with the TestCaseSourceAttribute), where a lot of test cases are loaded from CSV files or from Database. 
Now I realized while writing new tests, that all these testcase datasources, in all test classes of the test library are initialized prior to the execution of one test, even if my new test does not need any test case data source. This results in a huge waiting time when working test driven. Is there any way to restrict the initialization of datasources to the actual test class, or any other way except the TestCaseSourceAttribute to work with parameterized tests????

thanks and best regards,
peter

**Comment by CharliePoole** [comment]
Yes, that is the case. All your tests have to be created before we can run them. Creating the tests in this case is partly the job of your code that reads the database or file. Since NUnit treats each separate data point as a test, it does not even know how many tests there will be until your code has executed.

Currently, the best workaround is to minimize what you are doing in the methods referenced by your TestCaseSourceAttribute. My experience, however is that this is not easy to do when external sources actually determine the number of tests. If multiple tables are involved, you may get away with postponing some of the work until the test is run - for example, getting a list of user ids from one table and passing that to your actual test rather than immediately acquiring all the necessary data about the user from other tables or performing a join. Admittedly, this is more complicated and may not even work in some cases.

In the future, we plan to add what we call "dynamic tests", which would only be discovered at the tie the tests are run rather than when they are loaded. That would be the best solution for cases like yours and it's probably more like what you are used to with MsTest.

I have added this to the backlog.

**Comment by ptr1120** [comment]
thanks for the clarifying, quick reply. 

best regards,
peter

**Comment by CharliePoole** [comment]
I have added this issue to our backlog, as I said, so we will be sure to deal with it in the future. Closing it won't help get it done. :-)

**Comment by bryan-appeagle** [comment]
@ptr1120 I have somewhat of a workaround for this issue.  Inside of the TestCaseData method I am using a gate that is based on the current solution configuration.  For example if the configuration is Debug I have the method return a new TestCaseData with a .Ignore() message.  I then have another configuration, lets say Integration, which I use to run the integration tests and when I am using that configuration I load the test case data from the database.

You can do the gate's with #if statements or you could have the TestCaseData method call different methods that have been decorated with the .NET Conditional attribute.

Doing this the Debug configuration can be used by Developers for unit tests and then you can do integration tests in the other configuration on your CI server or manually.

Hope this helps you out.

**Comment by bryan-appeagle** [comment]
@CharliePoole I am having a similar issue and thought that by added the Category attribute to the tests and TestCaseSource attributes that it would prevent my long running TestCaseSource data initialization from running if I ignored tests in that category.

Based on the way I see NUnit working and your comment here I am guessing that NUnit processes all NUnit attributes like Test and TestCaseData for all classes in a test assembly and only uses the Category attribute to skip over the tests when the tests are run?

If that is how it's working couldn't you also check for a Category attribute while reflecting over the classes and methods in the assembly and if it contains an ignored category just skip over the class/method and treat it just like you treat classes and methods without NUnit attributes?

**Comment by CharliePoole** [comment]
@bryan-appeagle Yes, your understanding is correct. NUnit has to create all the tests so that they are available to be selected, whether by name, category or property. We do not restrict the tests in any way at load time.

We used to have a command-line option -fixture that caused only a specified class or namespace to be examined for tests. It was a bit hard to understand and turned out not to be much more efficient than simply waiting until runtime, so we removed it.

Since that time, it has become possible for users to do a lot of calculation in generating test case data, so there may be more time-saving possible in such cases.  The idea of dynamic tests is that the determination of what tests exist is postponed until runtime. So, for example, the cases in a fixture would not be created until that fixture was about to be run for the first time. I think this would give you the equivalent of what you are suggesting.

One caution: I'm struck at how easy it was to write that last paragraph, compared to how hard it's going to be to implement the feature. :-)

**Comment by bryan-appeagle** [comment]
Thanks for getting back to me so quickly.  Oh I am sure it will be as I've done some pretty complicated things with reflection over the years so I know how it is.  The workaround I've got for now is good as I can run our integration tests in a separate configuration so the test assembly that is generated for Unit Testing doesn't call the code to generate the data buy the configuration for Integration Testing does.

**Comment by rprouse** [comment]
Nobody has picked this up and nobody else has asked for this so closing.

**Comment by nzbart** [comment]
Just pointing out that there are plenty of people (like us) who would make heavy use of this feature, but "+1" comments are generally frowned upon. Please don't use the level of conversation to judge how useful a feature would be for end-users.

Having said that, we wouldn't have the time or expertise to develop the feature, so I can understand if it's too expensive to implement.
